qiskit/transpiler/passes/mapping/depth_bounded_mapper.py

```diff
@@ -0,0 +1,180 @@
+            return self.placement_cost(current_placement + place[0])
+
+        minimal_cost = sys.maxsize
+        while binops and matching:
+            # Find the cheapest binop to perform and minimize its cost.
+            min_min_placement = \
+                None  # type: Optional[Tuple[Placement[Reg, ArchNode], Tuple[Reg, Reg]]]
+            for binop in binops:
+                binop_map = {qarg: current_mapping[qarg] for qarg in binop.qargs}
+                # Try all matchings and find the minimum cost placement.
+                placements = (
+                    (Placement(binop_map, dict(zip(binop.qargs, node_ordering))), binop)
+                    for node0, node1 in matching
+                    for node_ordering in ((node0, node1), (node1, node0))
+                )  # type: Iterable[Tuple[Placement[Reg, ArchNode], DAGNode]]
+
+                min_placement = min(placements, key=placement_cost)
+
+                if min_min_placement is not None:
+                    min_min_placement = min(min_min_placement, min_placement, key=placement_cost)
+                else:
+                    min_min_placement = min_placement
+
+            if min_min_placement is None:
+                raise RuntimeError("The min_min_placement was not set.")
+
+            min_place_cost = placement_cost(min_min_placement)[0]
+            minimal_cost = min(minimal_cost, min_place_cost)  # Should only be set once.
+            logger.debug("Cost changing from %d → %d / %d",
+                         self.placement_cost(current_placement), min_place_cost, 2 * minimal_cost)
+
+            # If the placement cost of this placement exceeds the threshold,
+            # stop and go to next iteration.
+            if min_place_cost > 2 * minimal_cost:
+                logger.debug("Threshold exceeded!")
+                break
+
+            # Place the most expensive binop
+            current_placement += min_min_placement[0]
+            binops.remove(min_min_placement[1])
+            placed_gates += 1
+
+            # The nodes are now in use and can no longer be used for anything else.
+            remaining_arch.remove_nodes_from(min_min_placement[0].mapped_to.values())
+            # Both directions of the matching are now used.
+            matching.remove(frozenset(min_min_placement[0].mapped_to.values()))
+
+        logger.debug("Number of gates placed before: %d/%d", placed_gates, total_gates)
+
+        ###
+        # We now try to place the remaining gates on nodes
+        # such that they are closer to their destination.
+        # IDEA: apply some min-makespan instead of arbitrary ordering
+        ###
+        ordered_binops = list(binops)
+        random.shuffle(ordered_binops)  # Ensure random ordering
+        for binop in ordered_binops:
+            logger.debug("Placing %s", binop)
+            # Enumerate all nodes that do not exceed the cost threshold for both qargs
+            eligible_nodes = [[
```

Comment:
Can you unfold these nested list comprehension?

Reply:
What level of nesting is acceptable?

Reply:
This is an 8-line list comprehension.
```
Flat is better than nested.
Sparse is better than dense.
```
Maybe have a `eligible_nodes()` method?

Reply:
The unrolled alternative is
```
eligible_nodes = []
for qarg in binop.qargs:
    nodes = []
    for node in remaining_arch.nodes:
        # Rough filter of possible locations
        if self.distance[current_mapping[qarg]][node] <= 2 * minimal_cost
            # Exact filter
            and placement_cost((Placement({qarg: current_mapping[qarg]},
                {qarg: node}), binop.qargs))[0] <= 2 * minimal_cost:
            nodes.append(node)
    eligible_nodes.append(nodes)
```
not great either, whether it is in a function or not. I'll need to think about it for a bit.

Reply:
My problem is more about the list comprehension than with the nesting... I understand that line 141 collects a particular kind of node and unfold the code inline affect the reading flow. That's why I think a function can help.

the snippet `something <= 2 * minimal_cost` seems semantically relevant. Maybe it can also be a function.